### PR TITLE
Added creation for account account requisites and account activation.

### PR DESCRIPTION
**Pull request by OttoVT**
SIR-197
SIR-199

### CODE REVIEW

File: src/Brokerage.ApiContract/accounts.proto [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+message CreateAccountRequest {
+  int64 brokerAccountId = 1;
+
+  string requestId = 2;
+
+  string referenceId = 3;
+
+  string tenantId = 4;
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.ApiContract/accounts.proto]: Let's handle authorization in API for this request - remove tenant ID here and move the check to API

File: src/Brokerage.ApiContract/accounts.proto [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+  AccountStatus status = 3;
+
+  google.protobuf.Timestamp creationDateTime = 4;
+
+  google.protobuf.Timestamp activationDateTime = 5;
+
+  google.protobuf.Timestamp blockingDateTime = 6;
+
+  string referenceId = 7;
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.ApiContract/accounts.proto]: Should be StringValue

File: src/Brokerage.ApiContract/accounts.proto [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+  string tenantId = 8;
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.ApiContract/accounts.proto]: Not needed in the response

File: src/Brokerage.Common/Domain/AccountRequisites/TagType.cs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+﻿namespace Brokerage.Common.Domain.AccountRequisites
+{
+    public enum TagType
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Domain/AccountRequisites/TagType.cs]: We have this type in SDK

File: src/Brokerage.Common/Persistence/AccountRepository.cs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    {
+        private readonly DbContextOptionsBuilder<BrokerageContext> _dbContextOptionsBuilder;
+
+        public AccountRepository(DbContextOptionsBuilder<BrokerageContext> dbContextOptionsBuilder)
+        {
+            _dbContextOptionsBuilder = dbContextOptionsBuilder;
+        }
+
+
+        public async Task<Account> GetAsync(long accountId)
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Persistence/AccountRepository.cs]: It's a GetOrDefault according to the current impl

File: src/Brokerage.Common/Persistence/AccountRepository.cs [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+                return MapToDomain(newEntity);
+            }
+            catch (DbUpdateException e) //Check that request was already processed (by constraint)
+                when (e.InnerException is PostgresException pgEx &&
+                      pgEx.SqlState == "23505" &&
+                      pgEx.ConstraintName == "IX_Account_RequestId")
+            {
+                var entity = await context
+                    .Accounts
+                    .FirstOrDefaultAsync(x => x.RequestId == brokerAccount.RequestId);
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Persistence/AccountRepository.cs]: Why FirstOr**Default**?

File: src/Brokerage.Common/Persistence/AccountRepository.cs [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+                AccountStateEnum.Blocked => AccountState.Blocked,
+                AccountStateEnum.Creating => AccountState.Creating,
+                _ => throw new ArgumentOutOfRangeException($"{entity.State} is not processed")
+            };
+
+            var brokerAccount = Account.Restore(
+                entity.RequestId,
+                entity.AccountId,
+                entity.BrokerAccountId,
+                entity.BrokerAccount.TenantId,
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Persistence/AccountRepository.cs]: Don't do this.

File: src/Brokerage.Common/Persistence/AccountRepository.cs [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+            await using var context = new BrokerageContext(_dbContextOptionsBuilder.Options);
+
+            var newEntity = MapToEntity(brokerAccount);
+
+            context.Accounts.Add(newEntity);
+
+            try
+            {
+                await context.SaveChangesAsync();
+                await context.Entry(newEntity).Reference(s => s.BrokerAccount).LoadAsync();
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Persistence/AccountRepository.cs]: Avoid this

File: src/Brokerage.Common/Persistence/AccountRequisitesRepository.cs [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+                TagTypeEnum.Text => TagType.Text,
+                _ => throw new ArgumentOutOfRangeException(nameof(entity.TagType), entity.TagType.Value, null)
+
+            }) : null;
+
+            var brokerAccount = AccountRequisites.Restore(
+                entity.RequestId,
+                entity.Id,
+                entity.AccountId,
+                entity.Account.BrokerAccount.TenantId,
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Persistence/AccountRequisitesRepository.cs]: Don't do this

File: src/Brokerage.Common/Persistence/AccountRequisitesRepository.cs [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+
+        public async Task<AccountRequisites> GetAsync(string brokerAccountRequisitesId)
+        {
+            long.TryParse(brokerAccountRequisitesId, out var id);
+            await using var context = new BrokerageContext(_dbContextOptionsBuilder.Options);
+
+            var entity = await context
+                .AccountRequisites
+                .Include(x => x.Account)
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Persistence/AccountRequisitesRepository.cs]: Avoid this. Prefer to store denormalized data

File: src/Brokerage.Common/Persistence/AccountRequisitesRepository.cs [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+                .Include(x => x.Account.BrokerAccount)
+                .FirstOrDefaultAsync(x => x.Id == id);
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Persistence/AccountRequisitesRepository.cs]: Why OrDefault?

File: src/Brokerage.MessagingContract/BrokerAccountActivated.cs [hunk 12]
```diff
@@ -10,4 +8,11 @@ public class BrokerAccountActivated
 
         public DateTime ActivationDate { get; set; }
     }
+
+    public class AccountActivated
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.MessagingContract/BrokerAccountActivated.cs]: Move to separate file

File: src/Brokerage.Worker/MessageConsumers/FinalizeAccountCreationConsumer.cs [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+                    }
+
+                    requisites.Address = response.Response.Address;
+                    await _accountRequisitesRepository.UpdateAsync(requisites);
+                }
+
+            } while (true);
+
+
+            var account = await _accountRepository.GetAsync(message.BrokerAccountId);
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Worker/MessageConsumers/FinalizeAccountCreationConsumer.cs]: Should be AccountId here

File: src/Brokerage.Worker/MessageConsumers/FinalizeBrokerAccountCreationConsumer.cs [hunk 14]
```diff
@@ -89,9 +88,13 @@ public async Task Consume(ConsumeContext<FinalizeBrokerAccountCreation> context)
 
 
             var brokerAccount = await _brokerAccountRepository.GetAsync(message.BrokerAccountId);
-            brokerAccount.Activate();
 
-            await _brokerAccountRepository.UpdateAsync(brokerAccount);
+            if (brokerAccount.State == BrokerAccountState.Creating)
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Worker/MessageConsumers/FinalizeBrokerAccountCreationConsumer.cs]: We can even skip entire do/while above if the broker account is already not in Creating state and the same for account

File: src/Brokerage.ApiContract/Brokerage.ApiContract.csproj [hunk 15]
```diff
@@ -9,6 +9,10 @@
     <Product>Sirius</Product>
   </PropertyGroup>
 
+  <ItemGroup>
```
[the remark this answers is not in the document]
Comment (OttoVT) [reply src/Brokerage.ApiContract/Brokerage.ApiContract.csproj]: fixed

File: src/Brokerage.ApiContract/accounts.proto [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+service Accounts {
+  rpc Create (CreateAccountRequest) returns (CreateAccountResponse);
+}
+
+message CreateAccountRequest {
+  int64 brokerAccountId = 1;
+
+  string requestId = 2;
+
+  string referenceId = 3;
```
[the remark this answers is not in the document]
Comment (OttoVT) [reply src/Brokerage.ApiContract/accounts.proto]: fixed

File: src/Brokerage.ApiContract/accounts.proto (hunk 1 above)
Comment (OttoVT) [reply src/Brokerage.ApiContract/accounts.proto]: fixed

File: src/Brokerage.ApiContract/accounts.proto (hunk 2 above)
Comment (OttoVT) [reply src/Brokerage.ApiContract/accounts.proto]: fixed

File: src/Brokerage.ApiContract/accounts.proto (hunk 3 above)
Comment (OttoVT) [reply src/Brokerage.ApiContract/accounts.proto]: fixed

File: src/Brokerage.Common/Domain/AccountRequisites/TagType.cs (hunk 4 above)
Comment (OttoVT) [reply src/Brokerage.Common/Domain/AccountRequisites/TagType.cs]: fixed

File: src/Brokerage.Common/Persistence/AccountRepository.cs (hunk 5 above)
Comment (OttoVT) [reply src/Brokerage.Common/Persistence/AccountRepository.cs]: fixed

File: src/Brokerage.Common/Persistence/AccountRepository.cs (hunk 6 above)
Comment (OttoVT) [reply src/Brokerage.Common/Persistence/AccountRepository.cs]: fixed

File: src/Brokerage.Common/Persistence/AccountRepository.cs (hunk 7 above)
Comment (OttoVT) [reply src/Brokerage.Common/Persistence/AccountRepository.cs]: fixed

File: src/Brokerage.Common/Persistence/AccountRepository.cs (hunk 8 above)
Comment (OttoVT) [reply src/Brokerage.Common/Persistence/AccountRepository.cs]: fixed

File: src/Brokerage.Common/Persistence/AccountRequisitesRepository.cs (hunk 9 above)
Comment (OttoVT) [reply src/Brokerage.Common/Persistence/AccountRequisitesRepository.cs]: fixed

File: src/Brokerage.Common/Persistence/AccountRequisitesRepository.cs (hunk 10 above)
Comment (OttoVT) [reply src/Brokerage.Common/Persistence/AccountRequisitesRepository.cs]: fixed

File: src/Brokerage.Common/Persistence/AccountRequisitesRepository.cs (hunk 11 above)
Comment (OttoVT) [reply src/Brokerage.Common/Persistence/AccountRequisitesRepository.cs]: fixed

File: src/Brokerage.MessagingContract/BrokerAccountActivated.cs (hunk 12 above)
Comment (OttoVT) [reply src/Brokerage.MessagingContract/BrokerAccountActivated.cs]: fixed

File: src/Brokerage.Worker/MessageConsumers/FinalizeAccountCreationConsumer.cs (hunk 13 above)
Comment (OttoVT) [reply src/Brokerage.Worker/MessageConsumers/FinalizeAccountCreationConsumer.cs]: fixed

File: src/Brokerage.Common/Persistence/AccountRepository.cs [hunk 17]
```diff
@@ -18,7 +18,7 @@ public AccountRepository(DbContextOptionsBuilder<BrokerageContext> dbContextOpti
         }
 
 
-        public async Task<Account> GetAsync(long accountId)
+        public async Task<Account> GetOrDefaultAsync(long accountId)
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Persistence/AccountRepository.cs]: I meant it should be `Get` semantic implemented

File: src/Brokerage.Common/Persistence/AccountRequisitesRepository.cs (hunk 11 above)
Comment (KonstantinRyazantsev) [reply src/Brokerage.Common/Persistence/AccountRequisitesRepository.cs]: It's still FirstOrDefault

File: src/Brokerage.Worker/MessageConsumers/FinalizeAccountCreationConsumer.cs [hunk 18]
```diff
@@ -41,67 +41,71 @@ public async Task Consume(ConsumeContext<FinalizeAccountCreation> context)
...
+                    }
 
+                } while (true);
 
-            var account = await _accountRepository.GetAsync(message.BrokerAccountId);
-            account.Activate();
+                account.Activate();
 
-            await _accountRepository.UpdateAsync(account);
+                await context.Publish(new AccountActivated()
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Worker/MessageConsumers/FinalizeAccountCreationConsumer.cs]: The event should be published even if the account is not in Creating state

File: src/Brokerage.Worker/MessageConsumers/FinalizeBrokerAccountCreationConsumer.cs [hunk 19]
```diff
@@ -40,69 +40,72 @@ public async Task Consume(ConsumeContext<FinalizeBrokerAccountCreation> context)
...
+                        await _brokerAccountRequisitesRepository.UpdateAsync(requisites);
+                    }
+
+                } while (true);
 
-            if (brokerAccount.State == BrokerAccountState.Creating)
-            {
                 brokerAccount.Activate();
 
+                await context.Publish(new BrokerAccountActivated()
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Worker/MessageConsumers/FinalizeBrokerAccountCreationConsumer.cs]: The event should be published even if the account not in the creating state

File: src/Brokerage.Common/Persistence/AccountRequisitesRepository.cs (hunk 11 above)
Comment (OttoVT) [reply src/Brokerage.Common/Persistence/AccountRequisitesRepository.cs]: Fixed

File: src/Brokerage.Worker/MessageConsumers/FinalizeAccountCreationConsumer.cs [hunk 20]
```diff
@@ -41,70 +41,73 @@ public async Task Consume(ConsumeContext<FinalizeAccountCreation> context)
         {
             var message = context.Message;
             BlockchainId cursor = null;
-            var account = await _accountRepository.GetOrDefaultAsync(message.AccountId);
+            var account = await _accountRepository.GetAsync(message.AccountId);
 
             if (account == null)
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Worker/MessageConsumers/FinalizeAccountCreationConsumer.cs]: This check is not required since we use GetAsync above

**Pull request merged**